Add support for 'eetq' quantization format

### System Info

docker run --gpus all --shm-size 1g -p 8080:80 -v $volume:/data ghcr.io/predibase/lorax:latest --model-id $model --num-shard $num_shard --sharded true --max-total-tokens 16000 --max-input-length 15999 --max-batch-prefill-tokens 15999 --quantize eetq

### Information

- [X] Docker
- [ ] The CLI directly

### Tasks

- [ ] An officially supported command
- [ ] My own modifications

### Reproduction

error: invalid value 'eetq' for '--quantize <QUANTIZE>'
  [possible values: bitsandbytes, gptq]

### Expected behavior

it should run with "eetq" as a --quantize option . As TGI do support this quantised option

Looks like there was a build failure due to missing cutlass submodule:

```
#55 5.531 In file included from /usr/src/eetq/csrc/cutlass_kernels/fpA_intB_gemm_wrapper.cu:7:
#55 5.531 /usr/src/eetq/csrc/cutlass_kernels/fpA_intB_gemm.h:7:10: fatal error: cutlass/numeric_types.h: No such file or directory
#55 5.531     7 | #include "cutlass/numeric_types.h"
#55 5.531       |          ^~~~~~~~~~~~~~~~~~~~~~~~~
#55 5.531 compilation terminated.
#55 ...
```

@flozi00 will revert the commit for now.

--- a/.github/workflows/build.yaml
+++ b/.github/workflows/build.yaml
@@ -4,9 +4,10 @@ on:
   workflow_dispatch:
   push:
     branches:
-      - 'main'
+      - "main"
+      - "eetq"
     tags:
-      - 'v*'
+      - "v*"
 
 jobs:
   build-and-push-image:
@@ -41,8 +42,8 @@ jobs:
       - name: Install soci
         uses: lerentis/soci-installer@v1.0.1
         with:
-          soci-release: 'v0.4.0'
-      
+          soci-release: "v0.4.0"
+
       - name: Set up Docker Buildx
         uses: docker/setup-buildx-action@v2.10.0
 
@@ -51,7 +52,7 @@ jobs:
         with:
           config-inline: |
             version = 2
-            
+
             # persistent data location
             root = "/var/lib/kubelet/containerd"
 
@@ -62,11 +63,8 @@ jobs:
           images: |
             ghcr.io/predibase/lorax
           tags: |
-            type=semver,pattern={{version}}
-            type=semver,pattern={{major}}.{{minor}}
-            type=sha,prefix=,suffix=,format=short
-            type=raw,value=latest
-      
+            type=raw,value=eetq
+
       - name: Create a hash from tags
         env:
           tags: ${{ steps.meta.outputs.tags }}
@@ -93,7 +91,7 @@ jobs:
         uses: docker/build-push-action@v5
         with:
           context: .
-          file: ./Dockerfile  # Path to your Dockerfile
+          file: ./Dockerfile # Path to your Dockerfile
           push: false
           tags: ${{ steps.meta.outputs.tags }}
           outputs: type=oci,compression=gzip,dest=${{ steps.vars.outputs.image_path }}-${{ steps.vars.outputs.tag_hash }}.tar.gz
@@ -124,7 +122,7 @@ jobs:
             echo "Pushing $tag to GHCR"
             sudo ctr i push --user "${{ github.repository_owner }}:${{ secrets.GHCR_PAT }}" $tag
           done
-      
+
       - name: Create and push soci index
         env:
           tags: ${{ steps.meta.outputs.tags }}
@@ -151,4 +149,3 @@ jobs:
 
           # Delete the SHA image(s) from containerd store
           sudo ctr i rm $(sudo ctr i ls -q)
-  

--- a/.github/workflows/build.yaml
+++ b/.github/workflows/build.yaml
@@ -4,10 +4,9 @@ on:
   workflow_dispatch:
   push:
     branches:
-      - "main"
-      - "eetq"
+      - 'main'
     tags:
-      - "v*"
+      - 'v*'
 
 jobs:
   build-and-push-image:
@@ -42,8 +41,8 @@ jobs:
       - name: Install soci
         uses: lerentis/soci-installer@v1.0.1
         with:
-          soci-release: "v0.4.0"
-
+          soci-release: 'v0.4.0'
+      
       - name: Set up Docker Buildx
         uses: docker/setup-buildx-action@v2.10.0
 
@@ -52,7 +51,7 @@ jobs:
         with:
           config-inline: |
             version = 2
-
+            
             # persistent data location
             root = "/var/lib/kubelet/containerd"
 
@@ -63,8 +62,11 @@ jobs:
           images: |
             ghcr.io/predibase/lorax
           tags: |
-            type=raw,value=eetq
-
+            type=semver,pattern={{version}}
+            type=semver,pattern={{major}}.{{minor}}
+            type=sha,prefix=,suffix=,format=short
+            type=raw,value=latest
+      
       - name: Create a hash from tags
         env:
           tags: ${{ steps.meta.outputs.tags }}
@@ -91,7 +93,7 @@ jobs:
         uses: docker/build-push-action@v5
         with:
           context: .
-          file: ./Dockerfile # Path to your Dockerfile
+          file: ./Dockerfile  # Path to your Dockerfile
           push: false
           tags: ${{ steps.meta.outputs.tags }}
           outputs: type=oci,compression=gzip,dest=${{ steps.vars.outputs.image_path }}-${{ steps.vars.outputs.tag_hash }}.tar.gz
@@ -122,7 +124,7 @@ jobs:
             echo "Pushing $tag to GHCR"
             sudo ctr i push --user "${{ github.repository_owner }}:${{ secrets.GHCR_PAT }}" $tag
           done
-
+      
       - name: Create and push soci index
         env:
           tags: ${{ steps.meta.outputs.tags }}
@@ -149,3 +151,4 @@ jobs:
 
           # Delete the SHA image(s) from containerd store
           sudo ctr i rm $(sudo ctr i ls -q)
+  

--- a/Dockerfile
+++ b/Dockerfile
@@ -146,6 +146,13 @@ COPY server/punica_kernels/ .
 ENV TORCH_CUDA_ARCH_LIST="8.0;8.6+PTX"
 RUN python setup.py build
 
+# Build eetq kernels
+FROM kernel-builder as eetq-kernels-builder
+WORKDIR /usr/src
+COPY server/Makefile-eetq Makefile
+# Build specific version of transformers
+RUN TORCH_CUDA_ARCH_LIST="8.0;8.6+PTX" make build-eetq
+
 # LoRAX base image
 FROM nvidia/cuda:11.8.0-base-ubuntu20.04 as base
 
@@ -193,6 +200,9 @@ COPY --from=punica-builder /usr/src/build/lib.linux-x86_64-cpython-310 /opt/cond
 
 # Copy build artifacts from megablocks builder
 COPY --from=megablocks-kernels-builder /usr/src/megablocks/build/lib.linux-x86_64-cpython-310 /opt/conda/lib/python3.10/site-packages
+
+# Copy build artifacts from eetq builder
+COPY --from=eetq-kernels-builder /usr/src/eetq/build/lib.linux-x86_64-cpython-310 /opt/conda/lib/python3.10/site-packages
 
 # Install flash-attention dependencies
 RUN pip install einops --no-cache-dir

--- a/launcher/src/main.rs
+++ b/launcher/src/main.rs
@@ -53,6 +53,7 @@ impl std::fmt::Display for Quantization {
             }
             Quantization::Eetq => {
                 write!(f, "eetq")
+            }
             Quantization::Hqq_4bit => {
                 write!(f, "hqq-4bit")
             }

--- a/launcher/src/main.rs
+++ b/launcher/src/main.rs
@@ -26,6 +26,7 @@ enum Quantization {
     BitsandbytesFP4,
     Gptq,
     Awq,
+    Eetq,
 }
 
 impl std::fmt::Display for Quantization {
@@ -46,6 +47,9 @@ impl std::fmt::Display for Quantization {
             }
             Quantization::Awq => {
                 write!(f, "awq")
+            }
+            Quantization::Eetq => {
+                write!(f, "eetq")
             }
         }
     }

--- a/server/Makefile
+++ b/server/Makefile
@@ -2,6 +2,7 @@ include Makefile-flash-att
 include Makefile-flash-att-v2
 include Makefile-vllm
 include Makefile-megablocks
+include Makefile-eetq
 
 unit-tests:
 	pytest -s -vv -m "not private" tests

--- a/server/Makefile-eetq
+++ b/server/Makefile-eetq
@@ -7,7 +7,7 @@ eetq:
 
 build-eetq: eetq
 	cd eetq && git fetch && git checkout $(eetq_commit) && git submodule update --init --recursive
-	cd eetq && python setup.py build
+	cd eetq && MAX_JOBS=4 python setup.py build
 
 install-eetq: build-eetq
 	cd eetq && python setup.py install

--- a/server/Makefile-eetq
+++ b/server/Makefile-eetq
@@ -1,4 +1,4 @@
-eetq_commit := cc2fdb4637e03652ac264eaef44dd8492472de01
+eetq_commit := 323827dd471458a84e9c840f614e4592b157a4b1
 
 eetq:
     # Clone eetq
@@ -7,7 +7,7 @@ eetq:
 
 build-eetq: eetq
 	cd eetq && git fetch && git checkout $(eetq_commit) && git submodule update --init --recursive
-	cd eetq && MAX_JOBS=1 python setup.py build
+	cd eetq && python setup.py build
 
 install-eetq: build-eetq
 	cd eetq && python setup.py install

--- a/server/Makefile-eetq
+++ b/server/Makefile-eetq
@@ -1,4 +1,4 @@
-eetq_commit := 323827dd471458a84e9c840f614e4592b157a4b1
+eetq_commit := cc2fdb4637e03652ac264eaef44dd8492472de01 # 323827dd471458a84e9c840f614e4592b157a4b1
 
 eetq:
     # Clone eetq

--- a/server/Makefile-eetq
+++ b/server/Makefile-eetq
@@ -1,0 +1,13 @@
+eetq_commit := cc2fdb4637e03652ac264eaef44dd8492472de01
+
+eetq:
+    # Clone eetq
+	pip install packaging
+	git clone https://github.com/NetEase-FuXi/EETQ.git eetq
+
+build-eetq: eetq
+	cd eetq && git fetch && git checkout $(eetq_commit)
+	cd eetq && python setup.py build
+
+install-eetq: build-eetq
+	cd eetq && python setup.py install

--- a/server/Makefile-eetq
+++ b/server/Makefile-eetq
@@ -6,7 +6,7 @@ eetq:
 	git clone https://github.com/NetEase-FuXi/EETQ.git eetq
 
 build-eetq: eetq
-	cd eetq && git fetch && git checkout $(eetq_commit)
+	cd eetq && git fetch && git checkout $(eetq_commit) && git submodule update --init --recursive
 	cd eetq && python setup.py build
 
 install-eetq: build-eetq

--- a/server/Makefile-eetq
+++ b/server/Makefile-eetq
@@ -7,7 +7,7 @@ eetq:
 
 build-eetq: eetq
 	cd eetq && git fetch && git checkout $(eetq_commit) && git submodule update --init --recursive
-	cd eetq && MAX_JOBS=4 python setup.py build
+	cd eetq && MAX_JOBS=1 python setup.py build
 
 install-eetq: build-eetq
 	cd eetq && python setup.py install

--- a/server/lorax_server/cli.py
+++ b/server/lorax_server/cli.py
@@ -17,6 +17,7 @@ class Quantization(str, Enum):
     bitsandbytes_fp4 = "bitsandbytes-fp4"
     gptq = "gptq"
     awq = "awq"
+    eetq = "eetq"
 
 
 class Dtype(str, Enum):

--- a/server/lorax_server/utils/layers.py
+++ b/server/lorax_server/utils/layers.py
@@ -21,6 +21,14 @@ try:
 except ImportError:
     HAS_AWQ = False
 
+HAS_EETQ = False
+try:
+    from EETQ import quant_weights, w8_a16_gemm
+
+    HAS_EETQ = True
+except ImportError:
+    pass
+
 from accelerate import init_empty_weights
 
 from lorax_server.utils.gptq.quant_linear import QuantLinear
@@ -116,6 +124,78 @@ class FastConv1D(nn.Module):
         x = torch.addmm(self.bias, input.view(-1, input.size(-1)), self.weight)
         x = x.view(size_out)
         return x
+    
+
+class EETQLinear(nn.Module):
+    """
+    EETQLinear module applies quantized linear transformation to the input tensor.
+
+    Args:
+        weight (torch.Tensor): The weight tensor for the linear transformation.
+        bias (torch.Tensor): The bias tensor for the linear transformation.
+
+    Attributes:
+        weight (torch.Tensor): The weight tensor for the linear transformation.
+        scale (torch.Tensor): The scale tensor used for quantization.
+        bias (torch.Tensor): The bias tensor for the linear transformation.
+
+    """
+
+    def __init__(
+        self,
+        weight,
+        bias,
+    ) -> None:
+        super().__init__()
+        # Get the device where the weight tensor is currently stored.
+        device = weight.device
+
+        # Transpose the weight tensor and make a contiguous copy of it on the CPU.
+        # The contiguous() function is used to ensure that the tensor is stored in a contiguous block of memory,
+        # which can improve performance in some cases.
+        weight_transposed = torch.t(weight)
+        weight_contiguous = weight_transposed.contiguous()
+        weight_cpu = weight_contiguous.cpu()
+
+        # Quantize the weights. The quant_weights function is assumed to perform the quantization.
+        # The weights are quantized to int8 format, and the quantization is not performed in place (False).
+        weight_quantized, scale = quant_weights(weight_cpu, torch.int8, False)
+
+        # Move the quantized weights and the scale back to the original device (GPU if available).
+        # The cuda() function is used to move the tensors to the GPU.
+        self.weight = weight_quantized.cuda(device)
+        self.scale = scale.cuda(device)
+
+        # If a bias is present, move it to the GPU as well. If not, set the bias to None.
+        if bias is not None:
+            self.bias = bias.cuda(device)
+        else:
+            self.bias = None
+
+    def forward(self, input: torch.Tensor) -> torch.Tensor:
+            """
+            Performs the forward pass of the layer.
+
+            Args:
+                input (torch.Tensor): The input tensor.
+
+            Returns:
+                torch.Tensor: The output tensor.
+            """
+            # The function w8_a16_gemm performs a matrix multiplication operation between the input and the weight of the layer.
+            # The result is then scaled by a factor (self.scale).
+            gemm_output = w8_a16_gemm(input, self.weight, self.scale)
+
+            # If a bias is present (i.e., self.bias is not None), it is added to the output of the matrix multiplication.
+            # If a bias is not present (i.e., self.bias is None), the output of the matrix multiplication is returned as is.
+            if self.bias is not None:
+                final_output = gemm_output + self.bias
+            else:
+                final_output = gemm_output
+
+            # The final output is returned.
+            return final_output
+
 
 
 class Linear8bitLt(nn.Module):
@@ -243,6 +323,13 @@ def get_linear(weight, bias, quantize, fan_in_fan_out=False):
             bias,
             quant_type="fp4",
         )
+    elif quantize == "eetq":
+        if HAS_EETQ:
+            linear = EETQLinear(weight, bias)
+        else:
+            raise ImportError(
+                "Please install EETQ from https://github.com/NetEase-FuXi/EETQ"
+            )
     elif quantize == "gptq":
         try:
             qweight, qzeros, scales, g_idx, bits, groupsize, use_exllama = weight

